Петров Евгений

LeMarck opened the pull request:
@bratva, сори за профаканый дедлайн, времени не было(((((

LeMarck committed:
Add code

@bratva, сори за профаканый дедлайн, времени не было(((((

bratva commented:
Я проверю, но плохо, что после дедлайна.

bratva commented on lego.js:
```diff
@@ -2,7 +2,13 @@
-
+    var changedCollection = collection;
+    for (var func in arguments) {
+        if (func != '0') {
```
Для чего эта проверка?

bratva commented on lego.js:
```diff
@@ -17,11 +23,83 @@ module.exports.reverse = function () {
 
 // Оператор limit, который выбирает первые N записей
 module.exports.limit = function (n) {
```
Если я указал больше записей (например 1000), чем есть в коллекции - то возвращаются пустые элементы. 

Например в коллекции 10 записей, Я указал лимит 100. Мне должно вернуться 10 записей, а не 100

bratva commented on lego.js:
```diff
@@ -17,11 +23,83 @@ module.exports.reverse = function () {
+module.exports.filterIn = function (field, values) {
+    return function (collection) {
+        var changedCollection = [];
+        for (var element in collection) {
```
Посмотри на метод [Array.filter](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)

bratva commented on lego.js:
```diff
@@ -17,11 +23,83 @@ module.exports.reverse = function () {
+module.exports.sortBy = function (key, order) {
+    return function (collection) {
+        function sortedRule(first, second){
+            return first[key] - second[key];
```
`sort` ожидает что ты вернешь из обработчика `-1`, `0`, или `1`. [смотри](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/sort)

Ну и не работает сортировка по имени `"Вася" - "Петя" // NaN`

bratva commented:
Посмотри на методы
[Array.map](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/map)
[Array.forEach](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach)
[Array.some](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/some)

Постарайся их использовать, с ними будет красивее и легче.

bratva commented:
Будешь делать доп задание?

LeMarck committed:
Fixed

bratva commented on lego.js:
```diff
@@ -17,11 +23,76 @@ module.exports.reverse = function () {
 module.exports.limit = function (n) {
-    // Магия
+    // Магия.
+    return function (collection) {
```
Посмотри на метод `Array.slice` - он как раз делает то что нужно.

bratva commented on lego.js:
```diff
@@ -17,11 +23,76 @@ module.exports.reverse = function () {
+module.exports.select = function () {
+    var change = [].slice.call(arguments);
+    return function (collection) {
+        var changedCollection = collection.map(function(element){
```
``` js
return collection.map(function(element){
});
```

Можно сразу вернуть результат функции map, без присвоения его переменной `changedCollection`

bratva commented on lego.js:
```diff
@@ -17,11 +23,76 @@ module.exports.reverse = function () {
 
+module.exports.filterIn = function (field, values) {
+    return function (collection) {
+        var changedCollection = collection.filter(function(element){
```
Так же

``` js
return collection.filter(function(element){
})
```

bratva commented on lego.js:
```diff
@@ -17,11 +23,76 @@ module.exports.reverse = function () {
+module.exports.sortBy = function (key, order) {
+    return function (collection) {
+        function sortedRule(first, second){
+            return first[key] < second[key] ? -1 : first[key] > second[key] ? 1 : 0
```
не-не. Это же очень трудно читается. 
Давай лучше через `if else`

LeMarck committed:
Update

@bratva, ещё доп. задание скинул, которое забыл закомитить

LeMarck committed:
Update

@bratva, прогнал много раз OR - всё с ним ОК.
P.S.: в "книге" есть две Бориславы.

bratva commented:
`or` - так и не поправил. Будешь исправлять?

LeMarck commented:
@bratva, не могу понять что не так с OR
У меня все тесты проходит, если не сложно кинь вывод с дублями

bratva commented on lego.js:
```diff
@@ -17,11 +23,102 @@ module.exports.reverse = function () {
+        var changedCollection = [];
+        for (var element in change) {
+            var friends = change[element](collection);
+            changedCollection = changedCollection.concat(friends);
```
Ошибка тут.
Ты объеденяешь результаты выполнения всех фильтров.

``` js
collection = [
  {
      age: 33,
      gender: 'Мужской',
      name: 'Галактион',
      email: 'luisazamora@quilk.com',
      phone: '+7 (921) 505-3570',
      favoriteFruit: 'Картофель'
  },
  {
      age: 35,
      gender: 'Женский',
      name: 'Анна',
      email: 'roachpugh@mitroc.com',
      phone: '+7 (999) 539-2625',
      favoriteFruit: 'Яблоко'
  },
  {
      age: 21,
      gender: 'Мужской',
      name: 'Ефрем',
      email: 'danamcgee@permadyne.com',
      phone: '+7 (802) 526-2845',
      favoriteFruit: 'Картофель'
  }
];

var actual = lego.query(collection, lego.or(
    lego.filterEqual('gender', 'Мужской'),
    lego.filterIn('favoriteFruit', ['Картофель', 'Яблоко'])
));
```

Ожидаю 3 элемента -  а получаю 5.

``` js
[ { age: 33,
    gender: 'Мужской',
    name: 'Галактион',
    email: 'luisazamora@quilk.com',
    phone: '+7 (921) 505-3570',
    favoriteFruit: 'Картофель' },
  { age: 21,
    gender: 'Мужской',
    name: 'Ефрем',
    email: 'danamcgee@permadyne.com',
    phone: '+7 (802) 526-2845',
    favoriteFruit: 'Картофель' },
  { age: 33,
    gender: 'Мужской',
    name: 'Галактион',
    email: 'luisazamora@quilk.com',
    phone: '+7 (921) 505-3570',
    favoriteFruit: 'Картофель' },
  { age: 35,
    gender: 'Женский',
    name: 'Анна',
    email: 'roachpugh@mitroc.com',
    phone: '+7 (999) 539-2625',
    favoriteFruit: 'Яблоко' },
  { age: 21,
    gender: 'Мужской',
    name: 'Ефрем',
    email: 'danamcgee@permadyne.com',
    phone: '+7 (802) 526-2845',
    favoriteFruit: 'Картофель' } ]
```

Видно дубли, а надо

``` js
[ { age: 33,
    gender: 'Мужской',
    name: 'Галактион',
    email: 'luisazamora@quilk.com',
    phone: '+7 (921) 505-3570',
    favoriteFruit: 'Картофель' },
  { age: 35,
    gender: 'Женский',
    name: 'Анна',
    email: 'roachpugh@mitroc.com',
    phone: '+7 (999) 539-2625',
    favoriteFruit: 'Яблоко' },
  { age: 21,
    gender: 'Мужской',
    name: 'Ефрем',
    email: 'danamcgee@permadyne.com',
    phone: '+7 (802) 526-2845',
    favoriteFruit: 'Картофель' } ]
```

LeMarck committed:
fixed

bratva commented:
:star: молодец. Еще и `Set` применил.
